### PR TITLE
paraviewweb for 3D analysis

### PR DESCRIPTION
**Pull request by sanderegg**
---
name: **Add paraviewweb  Visualizer as a service in OSparc to use as a 3D analysis tool**
about: Paraviewweb Visualizer seems like a possible solution that could nicely connect #27. It is capable or reading rectilinear grid datasets which is the typical output of a LF solver. it has much of the built-in capabilities of paraview desktop (slice viewer, volume, mesh).

---
**Note:** Paraviewweb uses a running instance of paraview desktop to connect to and serves the results to any connected client. The tricky thing there is that docker does not provide with GUI capabilities so offscreen rendering is needed. Either through software rendering or using GPU on the docker worker.

- [X] Integration of paraviewweb visualiser inside of a docker using official master found in [repo](https://hub.docker.com/r/kitware/paraviewweb/tags/) - using OSMesa a software-based OpenGL implementation (current official release is 5.5 but it has a flaw when enabling local rendering mode)
- [x] Creation of simcore docker based on paraviewweb that adds:
    - a script (input-retriever.py) for downloading configuration and S3 input files
    - modify the apache 2 server by adding mod_cgi and configuration to call the python script on demand (through accessing /retrieve route)
    - retrieves the external port define by docker swarm when the service is started (external-port-retriever.py), needs to mount the docker node /var/sock to access the docker engine
- [x] docker-compose file that sets up S3/postgres to be able to test in local environment
- [x] Makefile/docker-compose similar to other services for building images, and later push to image registry
- [x] add service to the osparc interactive services
- [x] define interface (currently uses 2 input file-url)

**@odeimaiz @pcrespov @mguidon**

[waffle.io]:https://waffle.io/marketing-assets/documents/waffleio_cheatsheet_v1.pdf?utm_source=blog&utm_medium=cheatsheet-ctabutton&utm_campaign=cheatsheet

### CODE REVIEW

File: services/director/source/producer.py [hunk 1]
```diff
@@ -78,6 +90,16 @@ def __convert_labels_to_docker_runtime_parameters(service_runtime_parameters_lab
     _LOGGER.debug("Converted labels to docker runtime parameters: %s", runtime_params)
     return runtime_params
 
+def __get_service_entrypoint(service_boot_parameters_labels):
+    _LOGGER.debug("Getting service entrypoint")
+    for param in service_boot_parameters_labels:
+        if 'name' not in param or 'type' not in param or 'value' not in param:
```
Comment (mguidon) [inline services/director/source/producer.py]: Thats a bit silly.

File: services/director/source/producer.py [hunk 2]
```diff
@@ -149,6 +171,22 @@ def __get_docker_image_published_ports(service_id):
     _LOGGER.debug("Service %s publishes: %s ports", service_id, published_ports)
     return published_ports
 
+@tenacity.retry(wait=tenacity.wait_fixed(2), stop=tenacity.stop_after_attempt(3) or tenacity.stop_after_delay(10))
+def __pass_port_to_service(service, port, service_boot_parameters_labels):
+    for param in service_boot_parameters_labels:
+        if 'name' not in param or 'type' not in param or 'value' not in param:
+            pass
```
Comment (mguidon) [inline services/director/source/producer.py]: same here

File: services/dy-3dvis/Makefile [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+push_service_images:
+	${DOCKER} login masu.speag.com
+	${DOCKER} tag dy-3dvis_simcore-paraviewweb:latest ${SERVICE_NAME}:${SERVICES_VERSION}
+	${DOCKER} push ${SERVICE_NAME}:${SERVICES_VERSION}
+
+.env: .env-devel
+	$(info #####  $< is newer than $@ ####)
+	@diff -uN $@ $<
+	@false
```
Comment (mguidon) [inline services/dy-3dvis/Makefile]: Can/Should this be integrated into/called from the top level Makefile

File: services/dy-3dvis/docker-compose.devel.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      - postgres
+  #--------------------------------------------------------------------
+  minio:
+    image: minio/minio
+    environment:
+      - MINIO_ACCESS_KEY=${S3_ACCESS_KEY}
+      - MINIO_SECRET_KEY=${S3_SECRET_KEY}
+    ports:
+      - "9001:9000"
+    command: server /data
```
Comment (mguidon) [inline services/dy-3dvis/docker-compose.devel.yml]: Same here. Can we have command line arguments running only specific parts of the platform stack directly from the root docker-compose?

File: services/dy-3dvis/simcoreparaviewweb/config/apache.conf [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  SetEnv POSTGRES_ENDPOINT ${POSTGRES_ENDPOINT}
+  SetEnv S3_BUCKET_NAME ${S3_BUCKET_NAME}
+  SetEnv S3_SECRET_KEY ${S3_SECRET_KEY}
+  SetEnv S3_ACCESS_KEY ${S3_ACCESS_KEY}
+  SetEnv S3_ENDPOINT ${S3_ENDPOINT}
+  SetEnv SIMCORE_NODE_UUID ${SIMCORE_NODE_UUID}
+  SetEnv PYTHONPATH ${PYTHONPATH}
+  
+  Alias "/setport" "/home/root/src/setport.py"
+  Alias "/retrieve" "/home/root/src/input-retriever.py"
```
Comment (mguidon) [inline services/dy-3dvis/simcoreparaviewweb/config/apache.conf]: This looks cool. Is it secure? Not that I really care, just out of interest...

File: services/dy-3dvis/simcoreparaviewweb/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+minio==4.0.0
+networkx==2.1
+psycopg2-binary==2.7.4
```
Comment (mguidon) [inline services/dy-3dvis/simcoreparaviewweb/requirements.txt]: I remember having read that you should not use the psycopg2-binary but build from source instead, i.e. get the postgresql-dev packages from distro.

File: services/dy-3dvis/simcoreparaviewweb/scripts/apachePatch.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+echo "export POSTGRES_ENDPOINT=${POSTGRES_ENDPOINT}" >> /etc/environment
+echo "export POSTGRES_PASSWORD=${POSTGRES_PASSWORD}" >> /etc/environment
+echo "export POSTGRES_USER=${POSTGRES_USER}" >> /etc/environment
+echo "export POSTGRES_DB=${POSTGRES_DB}" >> /etc/environment
+echo "export PARAVIEW_INPUT_PATH=${PARAVIEW_INPUT_PATH}" >> /etc/environment
+
+echo "add aliases and script execution"
+sed -i -e '/CustomLog/r config/apache.conf' /etc/apache2/sites-available/001-pvw.conf
+
+a2enmod cgid
```
Comment (mguidon) [inline services/dy-3dvis/simcoreparaviewweb/scripts/apachePatch.sh]: Maybe comment what this is

File: services/director/source/producer.py (hunk 1 above)
Comment (sanderegg) [reply services/director/source/producer.py]: replaced by a function. don't know if this is less silly ;)
it now raise an exception. the idea is to detect whether the labels are in order or not.

File: services/dy-3dvis/docker-compose.devel.yml (hunk 4 above)
Comment (sanderegg) [reply services/dy-3dvis/docker-compose.devel.yml]: well basically, it is not part of the platform per se so I am not convinced this belongs with the root docker-compose file.
I use these devel compose to create a platform free testing environment that is faster to set-up (it automatically uploads whatever is necessary and reduces manually settings things). I think some of these things could be shared among services as it is probable that having a minimal S3,DB config will be necessary for a lot of services.

File: services/dy-3dvis/Makefile (hunk 3 above)
Comment (sanderegg) [reply services/dy-3dvis/Makefile]: see comment below

File: services/dy-3dvis/simcoreparaviewweb/config/apache.conf (hunk 5 above)
Comment (sanderegg) [reply services/dy-3dvis/simcoreparaviewweb/config/apache.conf]: secure? sure;) I guess as much as apache2 is...

File: services/dy-3dvis/simcoreparaviewweb/requirements.txt (hunk 6 above)
Comment (sanderegg) [reply services/dy-3dvis/simcoreparaviewweb/requirements.txt]: ok I'll have a look into this... I think I actually copy/pasted this from your stuff some time ago.. did you change that?

File: services/dy-3dvis/simcoreparaviewweb/scripts/apachePatch.sh (hunk 7 above)
Comment (sanderegg) [reply services/dy-3dvis/simcoreparaviewweb/scripts/apachePatch.sh]: done

**Pull request merged**